pwiz_tools/Shared/Common/SystemUtil/StreamEx.cs

```diff
@@ -0,0 +1,61 @@
+﻿/*
+ * Original author: Brendan MacLean <brendanx .at. u.washington.edu>,
+ *                  MacCoss Lab, Department of Genome Sciences, UW
+ *
+ * Copyright 2023 University of Washington - Seattle, WA
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using pwiz.Common.Properties;
+
+namespace pwiz.Common.SystemUtil
+{
+    public static class StreamEx
+    {
+        public static void TransferBytes(this Stream inStream, Stream outStream, long bytesToTransfer)
+        {
+            int bufferSize = (int)Math.Min(bytesToTransfer, 0x40000); // 256K;
+            inStream.TransferBytes(outStream, bytesToTransfer, new byte[bufferSize]);
+        }
+
+        public static void TransferBytes(this Stream inStream, Stream outStream, long bytesToTransfer, byte[] buffer)
+        {
+            long bytesTransferred = 0;
+            long bytesRemaining = bytesToTransfer;
+            int len;
+            while (bytesToTransfer > 0 && (len = inStream.Read(buffer, 0, (int)Math.Min(bytesRemaining, buffer.Length))) != 0)
+            {
+                outStream.Write(buffer, 0, len);
+                bytesRemaining -= len;
+                bytesTransferred += len;
+            }
+
+            if (bytesTransferred != bytesToTransfer)
+            {
+                throw new InvalidDataException(string.Format(
+                    @"Tried to transfer {0} bytes, but actual byte count transferred was {1}", bytesToTransfer,
+                    bytesTransferred));
+            }
+        }
+
+        public static void ReadOrThrow(this Stream stream, [In, Out] byte[] buffer, int offset, int count)
```

Comment:
I think this should be called something like "ReadComplete" or "ReadFully".
The problem is that "ReadOrThrow" does not give an indication of what would cause it to throw.

Also, in addition to this method, I kind of wish we also had a method which returned a new buffer of a particular length:
public static byte[] ReadByteCount(this Stream stream, int byteCount)
which returned a new buffer of the correct length.

Reply:
I don't know about the latter. In cases where we have an expected byte count, isn't it usually better to throw if we can't get that number of bytes because the file ends unexpectedly?

I did think first of "ReadAll" but that seemed too ambiguous. And I actually preferred being explicit that the main difference was that a failure to read the requested bytes would result in a thrown Exception. I am still not sure I think your proposed names are better.